File: 0200-number-of-islands/solution.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        """
        1. traverse the grid using bfs to find continue peices of land
        2. keep track of visited squares of grids in order not to double count
        """

        ans : int = 0
        rows, cols = len(grid), len(grid[0])
        visited :set[tuple[int]] = set()


        def bfs(row: int, col: int):
            q : deque = deque()
            q.append((row, col))
            visited.add((row, col))

            directions : list[tuple[int]] = [(1,0),(0,1),(-1,0),(0,-1)]  
            
            while q:
                row, col = q.popleft()
                for dx, dy in directions:
                    r, c = row + dx, col + dy

                    if (0 <= r < rows) and (0 <= c < cols) and (grid[r][c] == "1") and ((r, c) not in visited):
                        q.append((r, c))
                        visited.add((r, c))



        for r in range(rows):
            for c in range(cols):
                if (grid[r][c] == "1") and ((r, c) not in visited):
                    ans += 1
                    bfs(r, c)


        return ans

File: 0200-number-of-islands/solution.go
func numIslands(grid [][]byte) int {
	/*
	1. traverse the grid using bfs to find continue peices of land
	2. keep track of visited squares by setting their value to '2'
	*/
	ans := 0
	rows, cols := len(grid), len(grid[0])

	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			if grid[r][c] == '1' {
				ans += 1
				bfs(grid, r, c)
			}
		}
	}

	return ans
}


func bfs(grid [][]byte, row, col int) {
	q := [][2]int{}
	q = append(q, [2]int{row, col})
	grid[row][col] = '2'

	directions := [4][2]int{{1, 0}, {0, 1}, {-1, 0}, {0, -1}}

	for len(q) > 0 {
		rc := q[0]
		q = q[1:]
		for _, v := range directions {

			nr, nc := rc[0]+v[0], rc[1]+v[1]

			if 0 <= nr && nr < len(grid) && 0 <= nc && nc < len(grid[0]) && grid[nr][nc] == '1' {
				q = append(q, [2]int{nr, nc})
				grid[nr][nc] = '2'
			}

		}
	}

}

